Add X-Ray tracing to MIE stack

aburkleaux-amazon opened the pull request:
*Issue:* #251 

*Description of changes:*

Add X-Ray tracing to Media Insights Engine.

* Include xray packages in lambda layer
* Add xray packages to build virtualenv (for chalice)
* Use the lambda layer for all lambdas - this also solves code not loading in the console for some lambdas
* Add trace policies to all lambda IAM roles
* Add trace configuration to all lambdas
* Runtime patch of lambda packages to enable trace of boto3
* Add trace policies to step functions
* Enable xray in chalice APIs
* Add stack parameter to enable/disable X-Ray at deploy time

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

brandold commented on deployment/media-insights-stack.yaml:
```diff
@@ -225,6 +235,11 @@ Resources:
+                Action:
+                  - "xray:PutTraceSegments"
+                  - "xray:PutTelemetryRecords"
+                Resource: "*"
```
Does this have to be *? If it does, might be worth it to add a cfn_nag exception so we don't get flagged in the SB pipeline.

aburkleaux-amazon replied:
As far as I can tell, yes.  If you try using the IAM console to create this permission, it does not give any option to enter a specific resource.  It says "the actions you chose support all resources".

![image](https://user-images.githubusercontent.com/33331299/96290624-c71f5900-0f9b-11eb-9cd7-0f39a42dedd5.png)

aburkleaux-amazon replied:
If you are OK with this, can you point me to an example of the cfn nag I need to add

brandold replied:
WARN W11 is the message you'll need to add an exception for. I'll DM you an example.

brandold commented on deployment/media-insights-stack.yaml:
```diff
@@ -357,6 +372,12 @@ Resources:
+                Action:
+                  - "xray:PutTraceSegments"
+                  - "xray:PutTelemetryRecords"
+                Resource: "*"
```
Same as above comment

brandold commented on deployment/media-insights-stack.yaml:
```diff
@@ -398,6 +426,12 @@ Resources:
+                Action:
+                  - "xray:PutTraceSegments"
+                  - "xray:PutTelemetryRecords"
+                Resource: "*"
```
^^

brandold commented on deployment/media-insights-stack.yaml:
```diff
@@ -376,6 +397,13 @@ Resources:
+                  - "xray:PutTelemetryRecords"
+                  - "xray:GetSamplingRules"
+                  - "xray:GetSamplingTargets"
+                Resource: "*"
```
^^

brandold commented on source/dataplaneapi/external_resources.json:
```diff
@@ -52,6 +60,16 @@
+                      "xray:PutTelemetryRecords"
+                  ],
+                  "Resource": [
+                      "*"
```
Same as with other comments regarding * usage

brandold commented on source/workflow/requirements.txt:
```diff
@@ -1 +1 @@
-../../source/lib/MediaInsightsEngineLambdaHelper/dist/Media_Insights_Engine_Lambda_Helper-0.0.3-py3-none-any.whl
+
```
Was this just a stale artifact?

aburkleaux-amazon replied:
I added the lambda layer to this lambda so it doesn't need to package the lambda helper anymore.

brandold replied:
👍

brandold commented on source/workflowapi/external_resources.json:
```diff
@@ -145,6 +153,16 @@
+                      "xray:PutTelemetryRecords"
+                  ],
+                  "Resource": [
+                      "*"
```
Another * resource comment

brandold commented on source/workflowapi/.chalice/dev-app-policy.json:
```diff
@@ -41,6 +41,16 @@
+                "xray:PutTelemetryRecords"
+            ],
+            "Resource": [
+                "*"
```
Another * resource comment

brandold commented on source/operators/operator-library.yaml:
```diff
@@ -65,6 +65,11 @@ Resources:
+                Action:
+                  - "xray:PutTraceSegments"
+                  - "xray:PutTelemetryRecords"
+                Resource: "*"
```
Another * resource comment. Same for everything below.

brandold reviewed (commented):
My only items are (sorry for the multiple comments):

1. Usage of resource:* in all of the new IAM policies for xray. This is fine if there is no other option, but if we can either a) add a cfn_nag exception or b) update the resource to something other than * if possible

2. A couple deleted files - just want to double check we don't actually need them.

aburkleaux-amazon commented:
Regarding the deleted files... I found a number of files that I believe are dead code during this process.  I moved each of them aside and tested with a full build, deploy and manual execution test before I deleted them:

deployment/chalice-fix-inputs.py - is no longer needed since we can use the Chalice external resources feature now.
source/lib/MediaInsightsEngineWorkflow.py -  doesn't appear to be used anywhere
source/workflow/requirements.txt -  dependencies now come from the lambda layer
source/workflowapi/.chalice/apipolicy.json - not used anywhere, the chalice config references a different policy file
source/workflowapi/.chalice/config.json.layers  - not used anywhere, the chalice config references a different policy file

brandold commented:
👍
